File: September2022/C++/N-aryTreeLevelOrderTraversal.cpp
// https://leetcode.com/problems/n-ary-tree-level-order-traversal/
class Solution {
public:
    // TC: O(n)
    // SC: O(log(n))
    vector<vector<int>> levelOrder(Node* root) {
        vector<vector<int>> ans;
        if (root == NULL) {
            return ans;
        }
        
        queue<Node*> queue;
        queue.push(root);
        
        while (!queue.empty()) {
            int size = queue.size();
            vector<int> level;
            for (int i = 0; i < size; i++) {
                Node* node = queue.front();
                queue.pop();
                for (Node* it : node->children) {
                    if (it != NULL) {
                        queue.push(it);
                    }
                }
                level.push_back(node->val);
            }
            ans.push_back(level);
        }
        
        return ans;
    }
};

File: September2022/Java/NAryTreeLevelOrderTraversal.java

// @saorav21994
// TC : O(n)
// SC : O(n)

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null)
            return res;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        
        // standard bfs
        while (!queue.isEmpty()) {
            int lsize = queue.size();           // current level size
            List<Integer> sublist = new ArrayList<>();
            while (lsize-- > 0) {
                Node cur = queue.poll();
                int csize = cur.children.size();            // current node's children size
                for (int i = 0; i < csize; i++) {
                    queue.add(cur.children.get(i));
                }
                sublist.add(cur.val);
            }
            res.add(new ArrayList<>(sublist));
        }
        
        return res;
    }
}
